File: api/analyze.js
import { spawn } from 'child_process';
import path from 'path';

export default function handler(req, res) {
  if (req.method !== 'POST') {
    return res.status(405).json({ error: 'Method not allowed' });
  }

  const { koi_prad, koi_teq, koi_depth, koi_duration } = req.body || {};

  // Validate input parameters
  if (!koi_prad || !koi_teq || !koi_depth || !koi_duration) {
    return res.status(400).json({ error: 'All parameters are required' });
  }

  // Validate that all inputs are numbers
  const params = [koi_prad, koi_teq, koi_depth, koi_duration];
  if (params.some(param => isNaN(parseFloat(param)) || !isFinite(param))) {
    return res.status(400).json({ error: 'All parameters must be valid numbers' });
  }

  try {
    // Create a Python script that imports and uses the model implementation
    const pythonScript = `
import sys
import os
import json
import warnings

# Suppress sklearn warnings
warnings.filterwarnings('ignore')

# Add the public directory to Python path
sys.path.append('${path.join(process.cwd(), 'public').replace(/\\/g, '\\\\')}')

try:
    from model_implementation import predict_planet_status
    
    result = predict_planet_status(${koi_prad}, ${koi_teq}, ${koi_depth}, ${koi_duration})
    print("RESULT_START")
    print(json.dumps(result))
    print("RESULT_END")
    
except Exception as e:
    print("ERROR_START")
    print(str(e))
    print("ERROR_END")
`;

    // Execute Python script
    const python = spawn('python', ['-c', pythonScript]);
    
    let output = '';
    let errorOutput = '';

    python.stdout.on('data', (data) => {
      output += data.toString();
    });

    python.stderr.on('data', (data) => {
      errorOutput += data.toString();
    });

    python.on('close', (code) => {
      try {
        if (code !== 0) {
          console.error('Python script error:', errorOutput);
          return res.status(500).json({ 
            error: 'Python execution failed', 
            details: errorOutput 
          });
        }

        // Extract result from output
        const resultMatch = output.match(/RESULT_START\r?\n(.*?)\r?\nRESULT_END/s);
        const errorMatch = output.match(/ERROR_START\r?\n(.*?)\r?\nERROR_END/s);

        if (errorMatch) {
          return res.status(500).json({ 
            error: 'Model execution failed', 
            details: errorMatch[1].trim() 
          });
        }

        if (resultMatch) {
          const result = JSON.parse(resultMatch[1].trim());
          return res.status(200).json(result);
        } else {
          return res.status(500).json({ 
            error: 'Could not parse model output',
            output: output
          });
        }
      } catch (parseError) {
        console.error('Parse error:', parseError);
        return res.status(500).json({ 
          error: 'Failed to parse model results',
          details: parseError.message
        });
      }
    });

    python.on('error', (error) => {
      console.error('Python spawn error:', error);
      return res.status(500).json({ 
        error: 'Failed to execute Python script',
        details: error.message
      });
    });

  } catch (error) {
    console.error('API error:', error);
    return res.status(500).json({ 
      error: 'Internal server error',
      details: error.message
    });
  }
}

File: api/analyze.py
import json
import os
import sys
import warnings
from http.server import BaseHTTPRequestHandler
import urllib.parse

# Suppress sklearn warnings
warnings.filterwarnings('ignore')

# Add the current directory to Python path
current_dir = os.path.dirname(os.path.abspath(__file__))
sys.path.append(current_dir)

try:
    from model_implementation import predict_planet_status
except ImportError as e:
    print(f"Import error: {e}")
    predict_planet_status = None

class handler(BaseHTTPRequestHandler):
    def do_POST(self):
        try:
            # Check if model is available
            if predict_planet_status is None:
                self.send_response(500)
                self.send_header('Content-type', 'application/json')
                self.send_header('Access-Control-Allow-Origin', '*')
                self.end_headers()
                response = {
                    'error': 'Model not available',
                    'details': 'Could not import model_implementation'
                }
                self.wfile.write(json.dumps(response).encode())
                return

            # Get content length
            content_length = int(self.headers.get('Content-Length', 0))
            
            # Read the request body
            post_data = self.rfile.read(content_length)
            
            # Parse JSON data
            try:
                data = json.loads(post_data.decode('utf-8'))
            except json.JSONDecodeError:
                self.send_response(400)
                self.send_header('Content-type', 'application/json')
                self.send_header('Access-Control-Allow-Origin', '*')
                self.end_headers()
                response = {'error': 'Invalid JSON in request body'}
                self.wfile.write(json.dumps(response).encode())
                return

            # Extract parameters
            koi_prad = data.get('koi_prad')
            koi_teq = data.get('koi_teq')
            koi_depth = data.get('koi_depth')
            koi_duration = data.get('koi_duration')

            # Validate input parameters
            if not all([koi_prad, koi_teq, koi_depth, koi_duration]):
                self.send_response(400)
                self.send_header('Content-type', 'application/json')
                self.send_header('Access-Control-Allow-Origin', '*')
                self.end_headers()
                response = {'error': 'All parameters are required: koi_prad, koi_teq, koi_depth, koi_duration'}
                self.wfile.write(json.dumps(response).encode())
                return

            # Validate that all inputs are numbers
            try:
                koi_prad = float(koi_prad)
                koi_teq = float(koi_teq)
                koi_depth = float(koi_depth)
                koi_duration = float(koi_duration)
            except (ValueError, TypeError):
                self.send_response(400)
                self.send_header('Content-type', 'application/json')
                self.send_header('Access-Control-Allow-Origin', '*')
                self.end_headers()
                response = {'error': 'All parameters must be valid numbers'}
                self.wfile.write(json.dumps(response).encode())
                return

            # Make prediction
            try:
                result = predict_planet_status(koi_prad, koi_teq, koi_depth, koi_duration)
                
                # Send successful response
                self.send_response(200)
                self.send_header('Content-type', 'application/json')
                self.send_header('Access-Control-Allow-Origin', '*')
                self.end_headers()
                self.wfile.write(json.dumps(result).encode())
                
            except Exception as e:
                self.send_response(500)
                self.send_header('Content-type', 'application/json')
                self.send_header('Access-Control-Allow-Origin', '*')
                self.end_headers()
                response = {
                    'error': 'Model execution failed',
                    'details': str(e)
                }
                self.wfile.write(json.dumps(response).encode())

        except Exception as e:
            self.send_response(500)
            self.send_header('Content-type', 'application/json')
            self.send_header('Access-Control-Allow-Origin', '*')
            self.end_headers()
            response = {
                'error': 'Internal server error',
                'details': str(e)
            }
            self.wfile.write(json.dumps(response).encode())

    def do_OPTIONS(self):
        # Handle CORS preflight requests
        self.send_response(200)
        self.send_header('Access-Control-Allow-Origin', '*')
        self.send_header('Access-Control-Allow-Methods', 'POST, OPTIONS')
        self.send_header('Access-Control-Allow-Headers', 'Content-Type')
        self.end_headers()